books/kestrel/c/syntax/types.lisp

```diff
@@ -771,24 +908,71 @@
      @(':void') is compatible with @(':unknown'),
      as is @(':bool'),
      but @(':void') is <i>not</i> compatible with @(':bool')."))
-  (b* ((x (type-fix x))
-       (y (type-fix y)))
-    (or (equal x y)
-        (type-case x :unknown)
-        (type-case y :unknown)
-        (and (type-integerp x) (type-case y :enum))
-        (and (type-case x :enum) (type-integerp y))))
+  (or (type-case y :unknown)
+      (type-case
+        x
+        :unknown t
+        :pointer (type-case
+                   y
+                   :pointer (type-compatiblep x.to y.to)
+                   :otherwise nil)
+        :otherwise (or (equal (type-fix x) (type-fix y))
+                       (and (type-integerp x) (type-case y :enum))
+                       (and (type-case x :enum) (type-integerp y)))))
+  :measure (type-count x)
   :hooks (:fix)
 
   ///
 
   (defrule type-compatiblep-reflexive
     (type-compatiblep x x)
-    :enable type-compatiblep)
+    :induct t)
 
   (defrule type-compatiblep-symmetric
     (equal (type-compatiblep y x)
            (type-compatiblep x y))
+    :induct t))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(define type-composite ((x typep) (y typep))
+  :guard (type-compatiblep x y)
+  :returns (composite typep)
+  :short "Construct a composite @(see type) [C17:6.2.7/3]."
+  :long
+  (xdoc::topstring
+   (xdoc::p
+    "In our current approximate type system, the composite type is
+     the type of @('x') if the type of @('y') is unknown,
```

Comment:
Why "the type of x" vs. just "x", since x is a type? (Same for y.)

Reply:
I seem to have forgot that `x` and `y` were types when I wrote that :smile:. I'll replace "type of x/y" with "x/y".